server/webapp/WEB-INF/rails.new/app/controllers/environments_controller.rb

```diff
@@ -37,20 +37,20 @@ def new
   def create
     @environment.setConfigAttributes(params[:environment])
     if @environment.name().blank?
-      render_environment_create_error_with_message(@environment, l.string("ENVIRONMENT_NAME_REQUIRED"), 400)
+      render_environment_create_error_with_message(@environment, 'Environment name is required', 400)
       return
     end
 
     environment_config_service.createEnvironment(@environment, current_user, @result = HttpLocalizedOperationResult.new)
     render_environment_create_error_result(@environment)
-    redirect_with_flash(l.string("ADD_ENVIRONMENT_SUCCESS", [@environment.name()]), :action => :show, :name => @environment.name().to_s, :class => 'success') if @result.isSuccessful()
+    redirect_with_flash("Added environment '#{@environment.name}'", :action => :show, :name => @environment.name().to_s, :class => 'success') if @result.isSuccessful()
   end
 
   def update
     @environment = environment_config_service.getEnvironmentForEdit(params[:name])
     @environment.setConfigAttributes(params[:environment])
     if @environment.name().blank?
-      render_error_response l.string("ENVIRONMENT_NAME_REQUIRED"), 400, true
+      render_error_response 'Environment name is required', 400, true
```

Comment:
There are multiple places where a localized string could be used (Line 40 here). If we decide to change the message, we'd have to change it in multiple places. I am not sure how many such localize strings are there that are used in multiple places. I am assuming that you have done this for only certain localize strings and not for all.

Reply:
Yes, so far only for certain strings. I think what you're suggesting is to de-dup some of the strings, which I find useful. But given that we're moving over to SPAs, it's automatically taken care of. I don't particularly care for some of the old controllers and pages which have not changed in ages.

Reply:
Also — This is still a WIP. I'm basically in the middle of "inlining" all the constants in localize.xml (or as many as I can without making too much of a mess). I might need to extract some helper methods in a class if there's significant de-duplication that's needed and if it's worth it. So far, I've identified maybe a handful of messages that have a lot of duplication (in production code).